agent/internal/fluent.go

```diff
@@ -0,0 +1,450 @@
+	}
+
+	fluentArgs, fluentFiles, err := fluentConfig(opts)
+	if err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "failed to configure Fluent Bit")
+	}
+
+	// Decide what Docker network to use for Fluent Bit. If this agent is itself running inside Docker,
+	// use one of the networks that this container is using. If not, use the default network
+	// ("bridge").
+	dockerNet, err := getDockerNetwork(docker)
+	if err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "failed to get Docker network")
+	}
+	fluentDockerNet := dockerNet
+	if fluentDockerNet == "" {
+		fluentDockerNet = "bridge"
+	}
+	ctx.Log().Infof("running Fluent Bit on Docker network %q", fluentDockerNet)
+
+	createResponse, err := docker.ContainerCreate(
+		context.Background(),
+		&container.Config{
+			Image:        imageName,
+			Cmd:          fluentArgs,
+			ExposedPorts: nat.PortSet{exposedPort: struct{}{}},
+		},
+		&container.HostConfig{
+			AutoRemove:  true,
+			NetworkMode: container.NetworkMode(fluentDockerNet),
+			PortBindings: nat.PortMap{
+				// A port of 0 makes Docker automatically assign a free port, which we read back below.
+				exposedPort: []nat.PortBinding{{HostIP: "127.0.0.1", HostPort: "0"}},
+			},
+			Resources: container.Resources{
+				Memory:   1 << 30,
+				NanoCPUs: 1000000000,
+			},
+		},
+		nil,
+		containerName,
+	)
+	if err != nil {
+		return "", 0, 0, "", err
+	}
+
+	filesReader, _ := archive.ToIOReader(fluentFiles)
+
+	err = docker.CopyToContainer(context.Background(),
+		createResponse.ID,
+		"/",
+		filesReader,
+		types.CopyToContainerOptions{},
+	)
+	if err != nil {
+		return "", 0, 0, "", err
+	}
+
+	err = docker.ContainerStart(context.Background(), createResponse.ID, types.ContainerStartOptions{})
+	if err != nil {
+		return "", 0, 0, "", err
+	}
+
+	container, _ := docker.ContainerInspect(context.Background(), createResponse.ID)
+	portStr := container.NetworkSettings.Ports[exposedPort][0].HostPort
+	hostPort, _ := strconv.Atoi(portStr)
+	ctx.Log().Infof("Fluent Bit listening on host port %d", hostPort)
+
+	// We'll need to use either the container-internal or the host-visible address and port to connect
+	// to Fluent Bit, depending on whether this agent is running inside Docker or not.
+	addr := container.NetworkSettings.Networks[fluentDockerNet].IPAddress
+	port := fluentListenPort
+	if dockerNet == "" {
+		addr = "localhost"
+		port = hostPort
+	}
+
+	return addr, port, hostPort, createResponse.ID, nil
+}
+
+// getDockerNetwork returns the name of one of the Docker networks attached to the container that
+// this process is in. It returns an empty string if it does not seem to be running inside Docker or
+// is in Docker but on the host network, which should be treated equivalently.
+func getDockerNetwork(docker *client.Client) (string, error) {
+	f, err := os.Open("/proc/self/cgroup")
+	if err != nil {
+		return "", errors.Wrap(err, "error opening cgroup file")
+	}
```

Comment:
blocking: I think we need a switch case here to check if it's an `ErrNotFound`, otherwise outside of docker and not on linux (on mac) we'll fail here. I think general it's safe to assume that: IF we can't find this file OR this file doesn't contain the docker cgroup THEN we're not in a container.